storages/multipart_round_tripper.go

```diff
@@ -0,0 +1,208 @@
+
+		log.Debugf("Handling multi part upload, sending %s to %s, RequestID id %s",
+			request.URL.Path,
+			multiUploadBackend.Endpoint,
+			request.Context().Value(log.ContextreqIDKey))
+
+		response, requestError = multiUploadBackend.RoundTrip(request)
+
+		if requestError != nil {
+			log.Debugf("Error during multi part upload: %s", requestError)
+			return
+		}
+
+		if !isInitiateRequest(request) && isCompleteUploadResponseSuccessful(response) {
+			go multiPartRoundTripper.reportCompletionToMigrator(response, multiUploadBackend.Endpoint.String())
+		}
+
+		log.Debugf("Served multi part request, response code %d, status %s", response.StatusCode, response.Status)
+
+		return
+	}
+
+	return multiPartRoundTripper.fallBackRoundTripper.RoundTrip(request)
+}
+
+func (multiPartRoundTripper *MultiPartRoundTripper) pickBackend(objectPath string) *Backend {
+
+	for {
+
+		backendEndpoint, _ := multiPartRoundTripper.backendsRing.GetNode(objectPath)
+		backend, ok := multiPartRoundTripper.backendsRoundTrippers[backendEndpoint]
+
+		if !ok || backend.Maintenance {
+			continue
+		}
+
+		return backend
+	}
+}
+
+func (multiPartRoundTripper *MultiPartRoundTripper) isNotAbleToHandleMultiUpload() bool {
+	return len(multiPartRoundTripper.backendsRoundTrippers) < 1
+}
+
+func isMultiPartUploadRequest(request *http.Request) bool {
+	return isInitiateRequest(request) || containsUploadID(request)
+}
+
+func isInitiateRequest(request *http.Request) bool {
+	return strings.HasSuffix(request.URL.String(), "?uploads")
+}
+
+func containsUploadID(request *http.Request) bool {
+	return strings.Contains(request.URL.RawQuery, "uploadId=")
+}
+
+func isCompleteUploadResponseSuccessful(response *http.Response) bool {
+	return response.StatusCode == 200 &&
+		!strings.Contains(response.Request.URL.RawQuery, "partNumber=") &&
+		responseContainsCompleteUploadString(response)
+}
+
+func responseContainsCompleteUploadString(response *http.Response) bool {
+
+	responseBodyBytes, bodyReadError := ioutil.ReadAll(response.Body)
+
+	if bodyReadError != nil {
+
+		log.Debugf(
+			"Failed to read response body from CompleteMultipartUpload response for object %s, error: %s",
+			response.Request.URL, bodyReadError)
+
+		return false
+	}
+
+	response.Body = ioutil.NopCloser(bytes.NewBuffer(responseBodyBytes))
+
+	var completeMultipartUploadResult types.CompleteMultipartUploadResult
+
+	xmlParsingError := xml.Unmarshal(responseBodyBytes, &completeMultipartUploadResult)
+
+	if xmlParsingError != nil {
+
+		log.Debugf(
+			"Failed to parse body from CompleteMultipartUpload response for %s, error: %s",
+			response.Request.URL, xmlParsingError)
+
+		return false
+	}
+
+	log.Debugf("Successfully performed multipart upload to %s", completeMultipartUploadResult.Location)
+
+	return true
+}
+
+func (multiPartRoundTripper *MultiPartRoundTripper) reportCompletionToMigrator(response *http.Response, uploadedBackendName string) {
```

Comment:
AFAIK response.Request.Host == uploadedBackendName, so there is no need to pass extra argument